pkg/webhooks/trainjob_webhook.go

```diff
@@ -43,12 +46,31 @@ func setupWebhookForTrainJob(mgr ctrl.Manager, run map[string]runtime.Runtime) e
 
 var _ webhook.CustomValidator = (*TrainJobWebhook)(nil)
 
-func (w *TrainJobWebhook) ValidateCreate(context.Context, apiruntime.Object) (admission.Warnings, error) {
-	return nil, nil
+func (w *TrainJobWebhook) ValidateCreate(ctx context.Context, obj apiruntime.Object) (admission.Warnings, error) {
+	trainJob := obj.(*trainer.TrainJob)
+	log := ctrl.LoggerFrom(ctx).WithName("trainJob-webhook")
+	log.V(5).Info("Validating create", "TrainJob", klog.KObj(trainJob))
+	runtimeRefGK := runtime.RuntimeRefToRuntimeRegistryKey(trainJob.Spec.RuntimeRef)
+	runtime, ok := w.runtimes[runtimeRefGK]
+	if !ok {
+		return nil, fmt.Errorf("%s: %s", constants.UnsupportedRuntimeErrMsg, runtimeRefGK)
+	}
+	warnings, errorList := runtime.ValidateObjects(ctx, nil, trainJob)
+	return warnings, errorList.ToAggregate()
 }
 
-func (w *TrainJobWebhook) ValidateUpdate(context.Context, apiruntime.Object, apiruntime.Object) (admission.Warnings, error) {
-	return nil, nil
+func (w *TrainJobWebhook) ValidateUpdate(ctx context.Context, oldObj apiruntime.Object, newObj apiruntime.Object) (admission.Warnings, error) {
+	oldTrainJob := oldObj.(*trainer.TrainJob)
+	newTrainJob := newObj.(*trainer.TrainJob)
+	log := ctrl.LoggerFrom(ctx).WithName("trainJob-webhook")
+	log.V(5).Info("Validating update", "TrainJob", klog.KObj(newTrainJob))
+	runtimeRefGK := runtime.RuntimeRefToRuntimeRegistryKey(newTrainJob.Spec.RuntimeRef)
+	runtime, ok := w.runtimes[runtimeRefGK]
+	if !ok {
+		return nil, fmt.Errorf("%s: %s", constants.UnsupportedRuntimeErrMsg, runtimeRefGK)
+	}
+	warnings, errorList := runtime.ValidateObjects(ctx, oldTrainJob, newTrainJob)
```

Comment:
How do we use old TrainJob object in the ValidateObjects ?

Reply:
We currently don't use the values in the old object. Few such validations are taken care of with CEL expressions we added.